packages/use-storage/src/index.ts

```diff
@@ -0,0 +1,103 @@
+import {
+  useCallback,
+  useEffect,
+  useMemo,
+  useState,
+  useSyncExternalStore,
+} from 'react'
+import { flushSync } from 'react-dom'
+
+declare global {
+  interface WindowEventMap {
+    'event-storage': CustomEvent
+  }
+}
+
+const canUseDOM = !!(
+  typeof window !== 'undefined' &&
+  typeof window?.document !== 'undefined' &&
+  typeof window?.localStorage !== 'undefined' &&
+  typeof window?.sessionStorage !== 'undefined'
+)
+
+const subscribeStorage = (callback: () => void) => {
+  if (canUseDOM) {
+    window.addEventListener('storage', callback)
+    window.addEventListener('event-storage', callback)
+  }
+
+  return () => {
+    if (canUseDOM) {
+      window.removeEventListener('storage', callback)
+      window.removeEventListener('event-storage', callback)
+    }
+  }
+}
+
+const useStorage = <T>(
+  key: string,
+  options?: {
+    initialValue?: T
+    kind?: 'session' | 'local'
+  },
+): [T | null, (value: T | undefined) => void] => {
+  const storage = useMemo(
+    () =>
+      options?.kind === 'session' ? window.sessionStorage : window.localStorage,
+    [options?.kind],
+  )
+
+  const [localValue, setLocaleValue] = useState<string | null>(() =>
+    options?.initialValue ? JSON.stringify(options?.initialValue) : null,
+  )
+
+  const value = useSyncExternalStore(
+    subscribeStorage,
+    () => storage.getItem(key),
+    () => localValue,
+  )
+
+  useEffect(() => {
+    setLocaleValue(value)
+  }, [value])
+
+  const setValue = useCallback(
+    (val: T | undefined) => {
+      if (val) {
+        if (canUseDOM) {
+          flushSync(() => storage.setItem(key, JSON.stringify(val)))
+        }
+        setLocaleValue(JSON.stringify(val))
+      } else {
+        if (canUseDOM) {
+          flushSync(() => storage.removeItem(key))
```

Comment:
`storage.setItem()` / `storage.getItem()` doesn't trigger any React state update, so `flushSync` is useless (unless I'm missing something?)

Reply:
True, especially since we move to a subscription based state